[MLIR][Presburger] optimize bound computation by pruning orthogonal constraints

cxy-1993 opened the pull request:
IntegerRelation uses Fourier-Motzkin elimination and Gaussian elimination to simplify constraints. These methods may repeatedly perform calculations and elimination on irrelevant variables. Preemptively eliminating irrelevant variables and their associated constraints can speed up up the calculation process.

cxy-1993 committed:
[mlir][presburger] optimize bound computation by pruning orthogonal constraints

IntegerRelation uses Fourier-Motzkin elimination and Gaussian elimination to
simplify constraints. These methods may repeatedly perform calculations and
elimination on irrelevant variables. Preemptively eliminating irrelevant
variables and their associated constraints can speed up up the calculation process.

cxy-1993 commented:
> Hi @cxy-1993, sorry about the delay here. In "Prune constraints that are irrelevant to the target variable.", how is irrelevant defined?

In my opinion, the definition of variable irrelevance is as follows: If variables are considered as nodes and equations/inequalities as edges between nodes, a set of constraints forms an undirected graph. Variables in different connected components are irrelevant. This patch is about finding all nodes within the same connected component and then removing all nodes and related edges (equations and inequalities) that are not in the same connected component.

Superty commented:
I probably won't be able to look at this for a few more weeks, sorry. I would suggest adding some more documentation explaining what irrelevant means though. Is there a geometric interpretation of the definition? Maybe it makes sense to explain it with reference to Fourier motzkin?

The implementation could also use some more documentation. Thanks!

cxy-1993 commented:
> I probably won't be able to look at this for a few more weeks, sorry. I would suggest adding some more documentation explaining what irrelevant means though. Is there a geometric interpretation of the definition? Maybe it makes sense to explain it with reference to Fourier motzkin?
> 
> The implementation could also use some more documentation. Thanks!

I have now added detailed information regarding the function's mechanism and usage. Please let me know if you feel anything further is needed. Thank you for your time.

Superty commented on mlir/lib/Analysis/Presburger/IntegerRelation.cpp:
```diff
@@ -1723,12 +1724,78 @@ std::optional<DynamicAPInt> IntegerRelation::getConstantBoundOnDimSize(
+    } else {
+      unsigned currentCol = colStack.pop_back_val();
+      // Push all constrains that accociated to this variable to relatedRows
+      // and rowStack.
```
Please fix the spelling/grammar on this line (and above).

Superty commented on mlir/lib/Analysis/Presburger/IntegerRelation.cpp:
```diff
@@ -1723,12 +1724,78 @@ std::optional<DynamicAPInt> IntegerRelation::getConstantBoundOnDimSize(
   // Project to 'pos'.
+  pruneConstraints(pos);
   projectOut(0, pos);
+  pruneConstraints(0);
```
Can you write a comment here explaining why the second pruneConstraints might be helpful?

Superty commented on mlir/lib/Analysis/Presburger/IntegerRelation.cpp:
```diff
@@ -1723,12 +1724,78 @@ std::optional<DynamicAPInt> IntegerRelation::getConstantBoundOnDimSize(
+  }
+
+  // Prune all constraints not related to target variable.
+  for (int64_t constraintId = numConstraints - 1; constraintId >= 0;
```
You can just use an int for the constraintId.

Superty commented on mlir/include/mlir/Analysis/Presburger/IntegerRelation.h:
```diff
@@ -511,6 +511,31 @@ class IntegerRelation {
   inline void projectOut(unsigned pos) { return projectOut(pos, 1); }
 
+  /// The set of constraints (equations/inequalities) can be modeled as an
+  /// undirected graph where:
```
Can you write a 1-2 line summary at the top, so that people can quickly understand what the function does, while keeping the details below? You can write that the function removes some constraints that do not impose any bound on the specified variable.

superty reviewed (changes requested):
Thanks for your contribution and patience. I left a review!

(I removed the mention of "compilation time" from the title, I think we should reserve that phrase for the compilation time _of the library_)

cxy-1993 commented:
> Thanks for addressing the comments so quickly. Sorry, I should have explained what I meant a bit more clearly. I pushed a commit adding my proposed function ([00f3410](https://github.com/llvm/llvm-project/commit/00f3410719d090fe8aa77cc5ecc1a280c01fbf0d)). Can you refactor your code to use this? Thanks!

Excellent suggestion, I have made the modifications (including adding removeConstraint function).

cxy-1993 committed:
Merge branch 'main' into opt-intrelation

cxy-1993 committed:
[MLIR] [Presburger] fix compile error

cxy-1993 committed:
omit braces on simple if
